controller/service.go

```diff
@@ -47,76 +52,92 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
+	lbIPsIPFamily := ipfamily.Unknown
+	if len(lbIPs) == 0 {
 		c.clearServiceState(key, svc)
-		lbIP = nil
+	} else {
+		lbIPsIPFamily, err = ipfamily.ForAddressesIPs(lbIPs)
+		if err != nil {
+			level.Error(l).Log("event", "clearAssignment", "reason", "nolbIPsIPFamily", "msg", "Failed to retrieve lbIPs family")
+			c.client.Errorf(svc, "nolbIPsIPFamily", "Failed to retrieve LBIPs IPFamily for %q: %s", lbIPs, err)
+			return true
+		}
+		clusterIPsIPFamily, err := ipfamily.ForAddresses(svc.Spec.ClusterIPs)
+		if err != nil {
+			level.Error(l).Log("event", "clearAssignment", "reason", "noclusterIPsIPFamily", "msg", "Failed to retrieve clusterIPs family")
+			c.client.Errorf(svc, "noclusterIPsIPFamily", "Failed to retrieve ClusterIPs IPFamily for %q: %s", svc.Spec.ClusterIPs, err)
+			return true
+		}
+		// Clear the lbIP if it has a different ipFamily compared to the clusterIP.
+		// (this should not happen since the "ipFamily" of a service is immutable)
+		if lbIPsIPFamily != clusterIPsIPFamily {
+			c.clearServiceState(key, svc)
+			lbIPs = []net.IP{}
+		}
 	}
 
 	// It's possible the config mutated and the IP we have no longer
 	// makes sense. If so, clear it out and give the rest of the logic
 	// a chance to allocate again.
-	if lbIP != nil {
+	if len(lbIPs) != 0 {
 		// This assign is idempotent if the config is consistent,
 		// otherwise it'll fail and tell us why.
-		if err := c.ips.Assign(key, lbIP, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc)); err != nil {
-			level.Info(l).Log("event", "clearAssignment", "reason", "notAllowedByConfig", "msg", "current IP not allowed by config, clearing")
+		if err = c.ips.Assign(key, lbIPs, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc)); err != nil {
+			level.Info(l).Log("event", "clearAssignment", "error", err, "msg", "current IP not allowed by config, clearing")
 			c.clearServiceState(key, svc)
-			lbIP = nil
+			lbIPs = []net.IP{}
 		}
 
 		// The user might also have changed the pool annotation, and
 		// requested a different pool than the one that is currently
 		// allocated.
-		desiredPool := svc.Annotations["metallb.universe.tf/address-pool"]
-		if lbIP != nil && desiredPool != "" && c.ips.Pool(key) != desiredPool {
+		desiredPool := svc.Annotations[annotationAddressPool]
+		if len(lbIPs) != 0 && desiredPool != "" && c.ips.Pool(key) != desiredPool {
 			level.Info(l).Log("event", "clearAssignment", "reason", "differentPoolRequested", "msg", "user requested a different pool than the one currently assigned")
 			c.clearServiceState(key, svc)
-			lbIP = nil
+			lbIPs = []net.IP{}
+		}
+		// User set or changed the desired LB IP, nuke the
+		// state. allocateIP will pay attention to LoadBalancerIP and try
+		// to meet the user's demands.
+		// For dual-stack it will match against the 1st allocated IP.
+		if svc.Spec.LoadBalancerIP != "" && (lbIPsIPFamily == ipfamily.DualStack || svc.Spec.LoadBalancerIP != lbIPs[0].String()) {
```

Comment:
I'd split this in two:

svc.Spec.LoadBalancerIP != "" && lbIPsIPFamily == ipfamily.DualStack we fail.

if the ips are different we clear.